### PR TITLE
sets up the owner of the project

### LINKED ISSUE #1018: Start RBACGenerator implementation
TBD

child of #966

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:

makes sure that `ProjecGroup.Name` and `ProjectGroup.Group` with
proper values is attached to `User` object for the owner of the project.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1018 

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/controller/rbac/sync.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	// ensure owner
+	err = c.ensureProjectOwner(project)
+	return err
+}
+
+// ensureProjectOwner makes sure that the owner of the project is assign to "owners" group
+func (c *Controller) ensureProjectOwner(project *kubermaticv1.Project) error {
+	var owner *kubermaticv1.User
+	for _, ref := range project.OwnerReferences {
```
Comment (mrIncompetent) [inline api/pkg/controller/rbac/sync.go:31]: I think its not allowed (by convention) to have multiple owner references. Though the api allows it.
To get the owner i would prefer `ownerRef := metav1.GetControllerOf(node)`

Comment (p0lyn0mial) [reply api/pkg/controller/rbac/sync.go:31]: The user is the owner of the project not a controller.
`GetControllerOf` method gets the first `OnwerReference` that points to a controller.

Comment (p0lyn0mial) [reply api/pkg/controller/rbac/sync.go:31]: we could add assertion to check if there is only one owner ?

Comment (mrIncompetent) [reply api/pkg/controller/rbac/sync.go:31]: Why can the user not be considered controller?
Afaik `controller` in this context just means the parent object. Implying that when the parent gets deleted the garbage collection should delete all child objects as well.
Like with Deployment->Replicaset

Comment (p0lyn0mial) [reply api/pkg/controller/rbac/sync.go:31]: I think that GC will work regardless of whether `OwnerReference.Controller` is set or not.
I think `OwnerReference. blockOwnerDeletion` is more important for GC.

I have seen that `OnwerReference` is set usually by controllers by calling `NewControllerRef` method, e.g: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/sync.go#L197
which effectively sets also `OwnerReference.blockOwnerDeletion`

I have also pulled this extract from the docs
```
// If true, this reference points to the managing controller.
// +optional
Controller *bool `json:"controller,omitempty" protobuf:"varint,6,opt,name=controller"`
```
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L304


In addition 
```
// List of objects depended by this object. If ALL objects in the list have
// been deleted, this object will be garbage collected. If this object is managed by a controller,
// then an entry in this list will point to this controller, with the controller field set to true.
// There cannot be more than one managing controller.
// +optional
// +patchMergeKey=uid
// +patchStrategy=merge
OwnerReferences []OwnerReference `json:"ownerReferences,omitempty" patchStrategy:"merge" patchMergeKey:"uid" protobuf:"bytes,13,rep,name=ownerReferences"`
```

WDYT ?

Comment (mrIncompetent) [reply api/pkg/controller/rbac/sync.go:31]: sgtm

File: api/pkg/controller/rbac/sync.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+				return err
+			}
+		}
+	}
+	if owner == nil {
+		return fmt.Errorf("the given project %s doesn't have associated owner/user", project.Name)
+	}
+
+	for _, pg := range owner.Spec.Projects {
+		if pg.Name == project.Name {
```
Comment (p0lyn0mial) [inline api/pkg/controller/rbac/sync.go:44]: do we also want to check the group name ?

File: api/pkg/crd/kubermatic/v1/user.go [hunk 3]
```diff
@@ -43,6 +43,6 @@ type UserList struct {
 // stores the information about a project and a group that
 // a user belongs to
 type ProjectGroup struct {
-	ID    string `json:"id"`
+	Name  string `json:"name"`
```
Comment (p0lyn0mial) [inline api/pkg/crd/kubermatic/v1/user.go:46]: decided to change that to `Name` as this is more informative when getting the object.

**Review by mrIncompetent — APPROVED** [review]

File: api/pkg/controller/rbac/sync.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/controller/rbac/sync.go:44]: Should be fine i guess

**Pull request merged**